### PR TITLE
Death of swagger generation

### PR DESCRIPTION
**Pull request by defong**
### Context
Api killed swagger generated items

### Changes proposed in this pull request
Fixes/amended swagger related items

### Guidance to review
Everything should just work as expected

related to 
https://github.com/DFE-Digital/manage-courses-api/pull/211

### CODE REVIEW

**Comment by defong** [comment]
[Replace swagger-generated ApiClient with a handwritten class](https://trello.com/c/sy7jY6iC/532-replace-swagger-generated-apiclient-with-a-handwritten-class)

File: ManageCoursesUi.Tests/Controllers/OrganisationControllerTests.cs [hunk 1]
```diff
@@ -608,7 +606,7 @@ public void AboutPost_ModelState_WordCount()
                 .ReturnsAsync(institutionCourses);
 
 
-            var enrichmentModel = new InstitutionEnrichmentModel { AccreditingProviderEnrichments = new ObservableCollection<AccreditingProviderEnrichment> { } };
+            var enrichmentModel = new InstitutionEnrichmentModel { AccreditingProviderEnrichments = new List<AccreditingProviderEnrichment> { } };
```
Comment (timabell) [inline ManageCoursesUi.Tests/Controllers/OrganisationControllerTests.cs]: hurrah!

File: ManageCoursesUi.Tests/ViewModelHelperTests.cs [hunk 2]
```diff
@@ -86,22 +87,33 @@ public void TestGetSiteStatus(string status, string expectedResult)
...
+        [TestCase("b", "B", "Yes")]
+        [TestCase("b", "F", "Yes")]
+        [TestCase("b", "P", "Yes")]
+        [TestCase("b", "", "No")]
+
+        // These are probably wrong
+        [TestCase("", "B", "Yes")]
+        [TestCase("", "F", "Yes")]
+        [TestCase("", "P", "Yes")]
+        public void TestGetHasVacancies(string status, string vacStatus, string expectedResult)
```
Comment (timabell) [inline ManageCoursesUi.Tests/ViewModelHelperTests.cs]: how come the change of type of vacancy status got tied up with this PR?

File: src/ui/ActionFilters/McExceptionFilter.cs [hunk 3]
```diff
@@ -20,24 +20,26 @@ public McExceptionFilter(ILogger<McExceptionFilter> logger)
 
         public void OnException(ExceptionContext context)
         {
-            var swaggerException = context.Exception as SwaggerException;
+            var swaggerException = context.Exception as ManageCoursesApiException;
```
Comment (timabell) [inline src/ui/ActionFilters/McExceptionFilter.cs]: rename to match

File: src/ui/ManageApi.cs [hunk 4]
```diff
@@ -88,12 +90,12 @@ public async Task SaveCourseEnrichment(string instCode, string courseCode, Cours
         public async Task<SearchAndCompare.Domain.Models.Course> GetSearchAndCompareCourse(string instCode, string courseCode)
         {
             var result = await _apiClient.Publish_GetSearchAndCompareCourseAsync(instCode, courseCode);
-            //the result is an identical obect to the SearchAnCompareCourse that we want only it's an ApiClient version of it
-            //so we need to serialize/deserialize in order to get the required object to return
-            var jsonCourse = JsonConvert.SerializeObject(result, _jsonSerializerSettings);
-            SearchAndCompare.Domain.Models.Course deserializedCourse = JsonConvert.DeserializeObject<SearchAndCompare.Domain.Models.Course>(jsonCourse);
+            // //the result is an identical obect to the SearchAnCompareCourse that we want only it's an ApiClient version of it
+            // //so we need to serialize/deserialize in order to get the required object to return
```
Comment (timabell) [inline src/ui/ManageApi.cs]: redundant `//`

File: src/ui/ManageApi.cs [hunk 5]
```diff
@@ -88,12 +90,12 @@ public async Task SaveCourseEnrichment(string instCode, string courseCode, Cours
...
+            // var jsonCourse = JsonConvert.SerializeObject(result, _jsonSerializerSettings);
+            // SearchAndCompare.Domain.Models.Course deserializedCourse = JsonConvert.DeserializeObject<SearchAndCompare.Domain.Models.Course>(jsonCourse);
```
Comment (timabell) [inline src/ui/ManageApi.cs]: delete

File: src/ui/ManageCoursesUi.csproj [hunk 6]
```diff
@@ -28,7 +28,7 @@
   <Import Condition="Exists('$(ProjectDir)dev-sc-shared.targets')" Project="$(ProjectDir)dev-sc-shared.targets" />
 
   <ItemGroup Condition="Exists('$(ProjectDir)dev-mc-api.targets')==false">
-    <PackageReference Include="GovUk.Education.ManageCourses.ApiClient" Version="0.22.0.*" />
+    <PackageReference Include="GovUk.Education.ManageCourses.ApiClient" Version="0.22.0-death-to-swagger-generation.31" />
```
Comment (timabell) [inline src/ui/ManageCoursesUi.csproj]: don't forget to update this once the api is merged

File: src/ui/Startup.cs [hunk 7]
```diff
@@ -47,9 +49,23 @@ public void ConfigureServices(IServiceCollection services)
...
             var sharedAssembly = typeof(CourseDetailsViewComponent).GetTypeInfo().Assembly;
+
+
             services.AddMvc(options =>
                 options.Filters.Add(typeof(McExceptionFilter))
-            ).AddApplicationPart(sharedAssembly);
+            ).ConfigureApplicationPartManager(apm =>
+            {
+                var apiClientAssembly = typeof(GovUk.Education.ManageCourses.ApiClient.ManageCoursesApiClient).GetTypeInfo().Assembly.GetName().Name;
+            var api = typeof(GovUk.Education.ManageCourses.Api.Startup).GetTypeInfo().Assembly.GetName().Name;
```
Comment (timabell) [inline src/ui/Startup.cs]: reformat

File: src/ui/Startup.cs [hunk 8]
```diff
@@ -47,9 +49,23 @@ public void ConfigureServices(IServiceCollection services)
...
+                var dependentLibraries = apm.ApplicationParts.Where(x  => x.Name == api || x.Name == apiClientAssembly).ToList();
+
+                foreach (var item in dependentLibraries)
+                {
+                    apm.ApplicationParts.Remove(item);
```
Comment (timabell) [inline src/ui/Startup.cs]: perhaps a comment explaining why this is necessary

File: src/ui/package-lock.json [hunk 9]
```diff
@@ -10,7 +10,7 @@
       "integrity": "sha512-OfC2uemaknXr87bdLUkWog7nYuliM9Ij5HUcajsVcMCpQrcLmtxRbVFTIqmcSkSeYRBFBRxs2FiUqFJDLdiebA==",
       "dev": true,
       "requires": {
-        "@babel/highlight": "^7.0.0"
+        "@babel/highlight": "7.0.0"
```
Comment (timabell) [inline src/ui/package-lock.json]: I'd leave this file unchanged unless you feel like fixing anything that breaks

File: src/ui/Startup.cs (hunk 8 above)
Comment (defong) [reply src/ui/Startup.cs]: When you comment it out you know why.

File: src/ui/package-lock.json (hunk 9 above)
Comment (defong) [reply src/ui/package-lock.json]: Don't care its generated rubbish

File: ManageCoursesUi.Tests/ViewModelHelperTests.cs (hunk 2 above)
Comment (defong) [reply ManageCoursesUi.Tests/ViewModelHelperTests.cs]: The test vs the approach to make it pass changed as the underlying type has changed, vac was in master

Comment (defong) [reply ManageCoursesUi.Tests/ViewModelHelperTests.cs]: boy scout rulez appiled

File: ManageCoursesUi.Tests/Controllers/OrganisationControllerTests.cs (hunk 1 above)
Comment (defong) [reply ManageCoursesUi.Tests/Controllers/OrganisationControllerTests.cs]: what wrong? you wanted Observable`PainOfA`Collection back

File: src/ui/ManageCoursesUi.csproj (hunk 6 above)
Comment (defong) [reply src/ui/ManageCoursesUi.csproj]: When you approve it then this can be updated.

So you better work faster!

File: src/ui/Startup.cs (hunk 8 above)
Comment (defong) [reply src/ui/Startup.cs]: superfluous

Application Part Manager removes an Application Part that has the assembly with stated name

File: src/ui/ManageCoursesUi.csproj (hunk 6 above)
Comment (defong) [reply src/ui/ManageCoursesUi.csproj]: done

**Pull request merged**